### PR TITLE
Support Multiple Instances with Benchmarks

### PR DESCRIPTION
**Pull request by shawntabrizi**
The benchmarking pipeline was flawed in that we did not distinguish two pallets in the runtime that are using instances.

This adds that information throughout the pipeline, and then adds branching logic on the outputted filename.

If there are two pallets with the same name but different instances, the file output would be `pallet_name_instance_name.rs`, otherwise, the outputted file will be `pallet_name.rs`. Note that this logic only applies if a file name isn't specified. If the file name is specified, then we directly use that filename.

Tested with the Substrate node and `pallet_collective` for instancing, and `pallet_balances` without:

image

image

### CODE REVIEW

**Comment by thiolliere** [comment]
how that will work when location is not the instance ?
like in offences, session, and system in node/runtime:
```rust
			add_benchmark!(params, batches, pallet_offences, OffencesBench::<Runtime>);
			add_benchmark!(params, batches, pallet_proxy, Proxy);
			add_benchmark!(params, batches, pallet_scheduler, Scheduler);
			add_benchmark!(params, batches, pallet_session, SessionBench::<Runtime>);
			add_benchmark!(params, batches, pallet_staking, Staking);
			add_benchmark!(params, batches, frame_system, SystemBench::<Runtime>);
```

**Comment by shawntabrizi** [comment]
> maybe better if we add give another argument to add_benchmark which is the name of the file to be used ?

Yeah this is probably better.

File: utils/frame/benchmarking-cli/Cargo.toml [hunk 1]
```diff
@@ -23,11 +23,13 @@ sp-externalities = { version = "0.8.0", path = "../../../primitives/externalitie
...
 sp-runtime = { version = "2.0.0", path = "../../../primitives/runtime" }
 sp-state-machine = { version = "0.8.0", path = "../../../primitives/state-machine" }
-structopt = "0.3.8"
 codec = { version = "1.3.1", package = "parity-scale-codec" }
+structopt = "0.3.8"
 chrono = "0.4"
 serde = "1.0.116"
 handlebars = "3.5.0"
+Inflector = "0.11.4"
+
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/Cargo.toml:32]:
```suggestion
```

File: utils/frame/benchmarking-cli/src/writer.rs [hunk 2]
```diff
@@ -241,23 +245,30 @@ pub fn write_results(
...
 	let all_results = map_results(batches)?;
-	for (pallet, results) in all_results.into_iter() {
+	for ((pallet, instance), results) in all_results.iter() {
 		let mut file_path = path.clone();
 		// If a user only specified a directory...
 		if file_path.is_dir() {
-			// Create new file: "path/to/pallet_name.rs".
-			file_path.push(&pallet);
+			// Check if there might be multiple instances benchmarked.
+			if all_results.keys().find(|(p, i)| p == pallet && i != instance).is_some() {
```
Comment (bkchr) [inline utils/frame/benchmarking-cli/src/writer.rs:253]:
```suggestion
			if all_results.keys().any(|(p, i)| p == pallet && i != instance) {
```

File: frame/benchmarking/src/lib.rs [hunk 3]
```diff
@@ -1056,6 +1056,7 @@ macro_rules! impl_benchmark_test {
 macro_rules! add_benchmark {
 	( $params:ident, $batches:ident, $name:ident, $( $location:tt )* ) => (
 		let name_string = stringify!($name).as_bytes();
+		let instance_string = stringify!($( $location )* ).as_bytes();
```
Comment (bkchr) [inline frame/benchmarking/src/lib.rs:1059]:
```suggestion
		let instance_string = stringify!( $( $location )* ).as_bytes();
```

**Review by bkchr — DISMISSED** [review]

**Comment by shawntabrizi** [comment]
Decided to stick with the current syntax since it would prevent a breaking change, and is already really flexible with type aliases. I have added docs which make that clear.

**Review by thiolliere — APPROVED** [review]

**Comment by thiolliere** [comment]
maybe in the future we could do that is uses the name of the pallet given by construct_runtime, but that would be a breaking change so I'm fine with it.

**Review by kianenigma — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**